### PR TITLE
[stdlib] Give Sequence a top-level Element

### PR DESCRIPTION
**Pull request by airspeedswift**
Adds an `associatedtype Element` to `Sequence`, then constrains `Iterator.Element` to match it.

### CODE REVIEW

**Comment by airspeedswift** [comment]
OK I think the only ones left in the core stdlib are compiler crashers.

**Comment by airspeedswift** [comment]
@swift-ci Please benchmark

File: stdlib/public/core/Collection.swift [hunk 1]
```diff
@@ -60,11 +60,11 @@ public protocol _IndexableBase {
...
   // break a cyclic conformance/deduction that Swift can't handle.  We
-  // need something other than a Collection.Iterator.Element that can
+  // need something other than a Collection.Element that can
   // be used as IndexingIterator<T>'s Element.  Here we arrange for
   // the Collection itself to have an Element type that's deducible from
   // its subscript.  Ideally we'd like to constrain this Element to be the same
-  // as Collection.Iterator.Element (see below), but we have no way of
+  // as Collection.Element (see below), but we have no way of
   // expressing it today.
   associatedtype _Element
```
Comment (dabrahams) [inline stdlib/public/core/Collection.swift]: Can't we dump this now?

Comment (airspeedswift) [inline stdlib/public/core/Collection.swift]: No, seems we still need it, though I managed to ditch a handful of uses.

**Comment by dabrahams** [comment]
on Sat Apr 22 2017, Ben Cohen <notifications-AT-github.com> wrote:

> No, seems we still need it, though I managed to ditch a handful of uses.

OK.  You know Doug intends to make this change for us when he can
finally get it to work, yes?  Might be more efficient to let him pursue
these.

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by vmanot** [comment]
Is `associatedtype` really necessary? Would not `typealias` better reflect what we're trying to accomplish here, i.e. provide a substitute for `Iterator.Element`?

**Comment by timvermeulen** [comment]
@vmanot Conversely, using `associatedtype Element` seems more natural to me, since the `Element` type is usually the most relevant. If you were writing a `Sequence` protocol, wouldn't you start with `associatedtype Element` and later worry about how iteration is done?

**Comment by airspeedswift** [comment]
@timvermeulen right, that's my reasoning too. If you sat down with a blank screen but already had this feature in Swift, this is probably how you'd define it. In many cases both `Element` and `Iterator.Element` will usually be inferred. If you think about the common simple definitions of a new `Collection`, say, you would usually ignore `Iterator` and mostly focus on defining the `Element` in terms of what `subscript` returns.

**Comment by DougGregor** [comment]
This is fantastic!

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by shahmishal** [comment]
@swift-ci Please Test Source Compatibility

**Comment by airspeedswift** [comment]
@swift-ci Please test

**Comment by airspeedswift** [comment]
@swift-ci Please clean test linux platform

**Comment by dabrahams** [comment]
on Mon Apr 24 2017, Vatsal Manot <notifications-AT-github.com> wrote:

> Is `associatedtype` really necessary? Would not `typealias` better
> reflect what we're trying to accomplish here, i.e. provide a
> substitute for `Iterator.Element`?

It might, if typealiases in protocols worked for that purpose.

-- 
-Dave

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
Folded into #8990

**Pull request closed without merging**